packages/foundry/contracts/prediction-market/PredictionMarketHook.sol

```diff
@@ -0,0 +1,488 @@
+
+        // record the market particpation for the user and pool combination. This will allow the user to have 0 tx fees when 
+        // swapping in the pool to incentivize keeping the pool balanced via arbitrage. The latest timestamp value should be stored.
+        uint256 currentTimestamp = _predictionMarketPartipants[msg.sender][pool];
+
+        if(closedAtTimestamp > currentTimestamp){
+            _predictionMarketPartipants[msg.sender][pool] = closedAtTimestamp;
+        }
+
+        return updatedPosition;
+    }
+
+    /**
+     * @notice settle the given market
+     * @dev after markets are settled, participants can claim their payouts
+     *
+     * market must be past their end time into order to be claimed
+     * @param marketId the market to settle
+     * @return market the settled prediction market
+     *
+     */
+    function settle(bytes32 marketId) public returns (PredictionMarket memory market) {
+        address pool = markets[marketId].pool;
+        uint256 lastActivityBlock = _lastActivityBlockByPool[pool];
+
+        PredictionMarket memory settledMarket = markets.settle(marketId, lastActivityBlock, SETTLEMENT_WAITING_PERIOD, _vault);
+
+        _donate(settledMarket.pool, settledMarket.token0, settledMarket.fees);
+
+        return settledMarket;
+    }
+
+    /**
+     * @notice claim payouts from a settled market
+     * @dev the market must be settled in order to claim AND the sender must have a winning final position
+     * @param marketId the market to claim winnings from
+     * @return winnings the amount claimed to the msg.sender
+     */
+    function collect(bytes32 marketId) public returns (uint256 winnings){
+        // fetch the market and settle it try to settle it if it's still open
+        PredictionMarket memory market = markets[marketId];
+
+        if(!market.settled){
+            settle(marketId);
+        }
+
+        // determine the amount of funds to send to the user if any then mark the position as claimed
+        Position memory position = positions[marketId][msg.sender];
+
+        if(position.collected){
+            return 0;
+        }
+
+        uint256 userClaimAmount = position.getClaimableBalance(market);
+
+        positions[marketId][msg.sender].collected = true;
+
+        if(userClaimAmount > 0){
+            IERC20(market.token0).safeTransfer(msg.sender, userClaimAmount);
+        }
+
+        return userClaimAmount;
+    }
+
+    /**
+     * @notice swap a position holdings between sides 
+     * @dev bull -> bear or bear -> bull
+     *
+     * Fees are charged for each swap and maintained on the market for processing during settlement
+     * @param marketId Identifier of the market to swap
+     * @param from Side to swap from
+     * @param amountIn Amount to swap
+     * @return position Updated user position after the swap
+     */
+    function swap(
+        bytes32 marketId,
+        Side from,
```

Comment:
What if `from` is Side.Both? Can either revert if it's Both, or use a flag (e.g., `bool fromBull`).

Anything that's only called by users can be external, too.

Reply:
good catch, I added Both later on in development and it looks like i forgot to update swap to revert in that case